src/sagemaker_server.cc

```diff
@@ -628,69 +628,184 @@ SagemakerAPIServer::SageMakerMMEHandleInfer(
   }
 }
 
+TRITONSERVER_Error*
+SagemakerAPIServer::SageMakerMMECheckUnloadedModelIsUnavailable(
+    const char* model_name, bool* is_model_unavailable)
+{
+  /* Use the RepositoryIndex API to check if the model state has become
+  UNAVAILABLE i.e. model is no longer in the 'in-the-process-of' being
+  UNLOADED. Consequently, the reason field should be 'unloaded'.*/
+  TRITONSERVER_Message* server_model_index_message_ = nullptr;
+  uint32_t ready_flag = 0;  // value of 1 should be set if only the 'ready'
+                            // models are required from the index. In this case,
+                            // we need all models.
+  TRITONSERVER_ServerModelIndex(
+      server_.get(), ready_flag, &server_model_index_message_);
+
+  std::shared_ptr<TRITONSERVER_Message> managed_msg(
```

Comment:
Generally I'd recommend using `unique_ptr` rather than `shared_ptr` unless there's a specific reason `shared_ptr` is needed (and there's not, here).

Reply:
My concern was whether this is required due to multiple evhtp threads calling on the unload method. I can check more on this and modify at a later point in another PR, noted it...

@GuanLuo do you have any suggestion?

Reply:
The returned message is going to be unique to the caller, so there is no share of ownership here. `unique_ptr` will be better, except that syntax-wise it is a bit messy to attach custom deleter (in C++11).